File: go-cavebot/internal/combat/fighter.go
package combat

import "time"

type spellEntry struct {
	key      string
	cooldown float64
}

type SpellRotation struct {
	spells   []spellEntry
	lastUsed map[string]time.Time
}

func NewSpellRotation(keys []string, cooldowns []float64) *SpellRotation {
	spells := make([]spellEntry, len(keys))
	for i, k := range keys {
		cd := 0.0
		if i < len(cooldowns) {
			cd = cooldowns[i]
		}
		spells[i] = spellEntry{key: k, cooldown: cd}
	}
	return &SpellRotation{
		spells:   spells,
		lastUsed: make(map[string]time.Time),
	}
}

func (r *SpellRotation) NextReadySpell() (string, bool) {
	now := time.Now()
	for _, s := range r.spells {
		last := r.lastUsed[s.key]
		if now.Sub(last).Seconds() >= s.cooldown {
			return s.key, true
		}
	}
	return "", false
}

func (r *SpellRotation) MarkUsed(key string) {
	r.lastUsed[key] = time.Now()
}

// CastNext finds the next ready spell, calls pressFunc, and marks it used.
func (r *SpellRotation) CastNext(pressFunc func(string)) bool {
	key, ok := r.NextReadySpell()
	if !ok {
		return false
	}
	pressFunc(key)
	r.MarkUsed(key)
	return true
}

// Attack presses the attack key.
func Attack(attackKey string, pressFunc func(string)) {
	pressFunc(attackKey)
}

File: src/cavebot/combat/fighter.py
"""Attack and spell rotation via hotkeys."""

from __future__ import annotations

import time

from cavebot.core.input import press_key


class SpellRotation:
    def __init__(self, spell_keys: list[str], cooldowns: list[float]):
        self._spells = list(zip(spell_keys, cooldowns))
        self._last_used: dict[str, float] = {}

    def next_ready_spell(self) -> str | None:
        now = time.monotonic()
        for key, cooldown in self._spells:
            last = self._last_used.get(key, 0.0)
            if now - last >= cooldown:
                return key
        return None

    def mark_used(self, key: str) -> None:
        self._last_used[key] = time.monotonic()

    def cast_next(self) -> bool:
        key = self.next_ready_spell()
        if key is None:
            return False
        press_key(key)
        self.mark_used(key)
        return True


def attack(attack_key: str) -> None:
    press_key(attack_key)
